### PR TITLE
[Core] Fix status refresh for INIT cluster

### LINKED ISSUE #2490: [Core] Assertion error when refreshing status for INIT cluster
To reproduce:
```
sky launch -c existing-up-cluster
# ctrl-c when it is launching
sky status -r existing-up-cluster
sky launch -c existing-up-cluster
```

```
sky/backends/backend_utils.py", line 2024, in run_ray_status_to_check_ray_cluster_healthy
    assert handle.head_ssh_port is not None, handle
AssertionError: ResourceHandle(
        cluster_name=spot-dev-zhwu,
        cluster_name_on_cloud=spot-dev-zhwu,
        head_ip=34.135.151.192,
        stable_internal_external_ips=[('10.128.0.68', '34.135.151.192')],
        stable_ssh_ports=None,
        cluster_yaml=/Users/zhwu/.sky/generated/spot-dev-zhwu.yml, 
        launched_resources=1x GCP(n2-standard-8, image_id={'us-central1': 'projects/skypilot-managed-spot/global/machineImages/spot-dev-image'}), 
        docker_user=None,
        tpu_create_script=None, 
        tpu_delete_script=None)
```

**This is a serious problem, as once the cluster enter this state, none of our clis can operate on the cluster to get it out of the state.**

### PR DESCRIPTION
**Pull request by Michaelvll**
Fixes #2490 

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Any manual or new tests for this PR (please specify below)
  - [x] reproducible code in #2490
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

File: sky/backends/backend_utils.py [hunk 1]
```diff
@@ -2014,17 +2014,24 @@ def run_ray_status_to_check_ray_cluster_healthy() -> bool:
...
                     reason=exceptions.FetchIPError.Reason.HEAD)
 
+            if handle.head_ssh_port is None:
+                # Refresh the ssh ports. It is ok to refresh as it is fast.
+                handle.external_ssh_ports()
+                if handle.head_ssh_port is None:
+                    logger.debug(
+                        f'Refreshing status ({cluster_name!r}): failed '
+                        f'to get the ssh ports. Handle: {handle}')
+                    raise exceptions.FetchIPError(
```
Comment (cblmemo) [inline sky/backends/backend_utils.py:2028]: Shall we use a new exception named `FetchPortError` instead?

Comment (Michaelvll) [reply sky/backends/backend_utils.py:2028]: Changed to `RuntimeError`. Wdyt?

File: sky/backends/backend_utils.py [hunk 2]
```diff
@@ -2014,14 +2014,22 @@ def run_ray_status_to_check_ray_cluster_healthy() -> bool:
...
             if external_ips is None or len(external_ips) == 0:
                 logger.debug(f'Refreshing status ({cluster_name!r}): No cached '
-                             f'IPs found. External IPs: {external_ips}')
+                             f'IPs found. Handle: {handle}')
                 raise exceptions.FetchIPError(
                     reason=exceptions.FetchIPError.Reason.HEAD)
 
+            if handle.head_ssh_port is None:
+                # Refresh the ssh ports. It is ok to refresh as it is fast.
+                handle.external_ssh_ports()
```
Comment (cblmemo) [inline sky/backends/backend_utils.py:2023]: Shall we use `update_ssh_ports` for better readability?

Comment (cblmemo) [inline sky/backends/backend_utils.py:2023]: Shall we use `update_ssh_ports` for better readability?

File: sky/backends/cloud_vm_ray_backend.py [hunk 3]
```diff
@@ -1543,10 +1544,11 @@ def _retry_zones(
                 launched_resources=launched_resources,
                 tpu_create_script=config_dict.get('tpu-create-script'),
                 tpu_delete_script=config_dict.get('tpu-delete-script'),
-                # Use the previous cluster's IPs if available to optimize
-                # the case where the cluster is restarted, i.e., no need to
-                # query the IPs from the cloud provider.
-                stable_internal_external_ips=prev_cluster_ips)
+                # Use the previous cluster's IPs and ports if available to
+                # optimize the case where the cluster is restarted, i.e., no
+                # need to query the IPs from the cloud provider.
```
Comment (cblmemo) [inline sky/backends/cloud_vm_ray_backend.py:1549]:
```suggestion
                # need to query the IPs and ports from the cloud provider.
```

File: sky/backends/backend_utils.py (hunk 1 above)
Comment (cblmemo) [reply sky/backends/backend_utils.py]: Sounds good!

**Review by cblmemo — APPROVED** [review]
Just tested and it works great for me! Left several nits ;)

**Pull request merged**